components/feature/syncedtabs/src/main/java/mozilla/components/feature/syncedtabs/SyncedTabsFeature.kt

```diff
@@ -5,80 +5,50 @@
+import mozilla.components.feature.syncedtabs.internal.SyncedTabsController
+import mozilla.components.feature.syncedtabs.internal.SyncedTabsInteractor
+import mozilla.components.feature.syncedtabs.internal.SyncedTabsPresenter
+import mozilla.components.feature.syncedtabs.view.SyncedTabsView
 import mozilla.components.service.fxa.manager.FxaAccountManager
-import mozilla.components.service.fxa.manager.ext.withConstellation
-import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import kotlin.coroutines.CoroutineContext
 
 /**
- * A feature that listens to the [BrowserStore] changes to update the local remote tabs state
- * in [RemoteTabsStorage].
+ * Feature implementation that will keep a [SyncedTabsView] notified with other synced device tabs for
+ * the Firefox Sync account.
+ *
+ * @param storage The synced tabs storage that stores the current device's and remote device tabs.
+ * @param accountManager Firefox Account Manager that holds a Firefox Sync account.
+ * @param view An implementor of [SyncedTabsView] that will be notified of changes.
+ * @param lifecycleOwner Android Lifecycle Owner to bind observers onto.
+ * @param coroutineContext A coroutine context that can be used to perform work off the main thread.
+ * @param onTabClicked Invoked when a tab is selected by the user on the [SyncedTabsView].
  */
-@ExperimentalCoroutinesApi
 class SyncedTabsFeature(
-    private val accountManager: FxaAccountManager,
-    private val store: BrowserStore,
-    private val tabsStorage: RemoteTabsStorage = RemoteTabsStorage()
-) {
-    private var scope: CoroutineScope? = null
-
-    /**
-     * Start listening to browser store changes.
-     */
-    fun start() {
-        scope = store.flowScoped { flow ->
-            flow.ifChanged { it.tabs }.map { state ->
-                // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5178
-                val lastUsed = 0L
-                // TO-DO: https://github.com/mozilla-mobile/android-components/issues/5179
-                val iconUrl = null
-                state.tabs.filter { !it.content.private }.map { tab ->
-                    // TO-DO: https://github.com/mozilla-mobile/android-components/issues/1340
-                    val history = listOf(TabEntry(tab.content.title, tab.content.url, iconUrl))
-                    Tab(history, 0, lastUsed)
-                }
-            }.collect { tabs ->
-                tabsStorage.store(tabs)
-            }
-        }
-    }
-
-    /**
-     * Stop listening to browser store changes.
-     */
-    fun stop() {
-        scope?.cancel()
-    }
+    storage: SyncedTabsStorage,
+    accountManager: FxaAccountManager,
+    view: SyncedTabsView,
+    lifecycleOwner: LifecycleOwner,
+    coroutineContext: CoroutineContext = Dispatchers.IO,
+    onTabClicked: (Tab) -> Unit
+) : LifecycleAwareFeature {
+    @VisibleForTesting
```

Comment:
I would just lift these fields into the constructor. This way you can override them in your test too, right?

Reply:
Yeah I could, although that's not something that app should we aware of. Those are part of our internals for the feature.

I couldn't find a better alternative for this, so I opted for doing the same thing [as FindInPageFeature](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/findinpage/src/main/java/mozilla/components/feature/findinpage/FindInPageFeature.kt). Maybe there is a better way?

Reply:
Spoke offline. Made a public API for the interactor/presenter/controller that can be overridden if needed.